h2o-mapreduce-generic/src/main/java/water/hadoop/FileSystemBasedClouding.java

```
@@ -0,0 +1,140 @@
+    _fs = FileSystem.get(conf);
+  }
+
+  @Override
+  public void notifyAboutCloudSize(InetAddress ip, int port, InetAddress leaderIp, int leaderPort, int size) {
+    if (size != _n)
+      return;
+    Path path = toNodePath(ip, port, "leader");
+    try {
+      writeFile(path, h2oUri(leaderIp, leaderPort).toString());
+    } catch (IOException e) {
+      e.printStackTrace();
+      exit(162);
+    }
+  }
+
+  @Override
+  public void notifyAboutEmbeddedWebServerIpPort(InetAddress ip, int port) {
+    _self_ip = ip;
+    _self_port = port;
+
+    Path nodePath = toNodePath(ip, port, "node");
+    try {
+      writeFile(nodePath, h2oUri(ip, port).toString());
+    } catch (IOException e) {
+      e.printStackTrace();
+      exit(160);
+    }
+  }
+
+  @Override
+  public boolean providesFlatfile() {
+    return true;
+  }
+
+  @Override
+  public String fetchFlatfile() throws Exception {
+    while (_nodes.size() < _n) {
+      for (FileStatus fs : _fs.listStatus(_path, new NewNodesPathFilter())) {
+        String name = fs.getPath().getName();
+        if (name.endsWith(".exit")) {
+          exit(161);
+          return null;
+        }
+        try (InputStream is = _fs.open(fs.getPath());
+             BufferedReader br = new BufferedReader(new InputStreamReader(is))) {
+          String node = br.readLine();
+          _nodes.put(name, URI.create(node));
+        }
+      }
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+    }
+    StringBuilder sb = new StringBuilder();
+    for (URI node : _nodes.values()) {
+      sb.append(node.getHost()).append(':').append(node.getPort());
+      sb.append('\n');
+    }
+    return sb.toString();
+  }
+  
+  @Override
+  public void exit(int status) {
+    Path exitPath = toNodePath(_self_ip, _self_port, "exit");
+    try {
+      writeFile(exitPath, String.valueOf(status));
+    } catch (Exception e) {
+      e.printStackTrace();
```

eating exceptions is always bad, I would wrap in RuntimeException and rethrow

---

that would defeat the purpose, the node would not actually shutdown

this is safe to do in Hadoop context